sw/device/bsp/w25q/w25q.c

```diff
@@ -0,0 +1,1247 @@
+*/
+static void page_write(uint32_t addr, uint8_t *data, uint32_t length, uint8_t quad, uint8_t dma);
+
+/**
+ * @brief Copy length bytes from data to the SPI TX FIFO, using DMA.
+ * 
+ * @param data pointer to the data buffer.
+ * @param length number of bytes to copy.
+*/
+static uint8_t dma_send_toflash(uint8_t *data, uint32_t length);
+
+/**
+ * @brief Enable flash write.
+ * 
+ * It sets the WEL bit in the flash status register.
+ * Every action that require the WEL to be set is goig to automatically
+ * clear it.
+*/
+static void flash_write_enable(void);
+
+/**
+ * @brief Performs sanity checks on the input parameters.
+ * 
+ * Checks if the address is valid, the data pointer is not NULL
+ * and the length is not 0.
+ * 
+ * @param addr 24-bit address.
+ * @param data pointer to the data buffer.
+ * @param length number of bytes to read/write.
+ * @return 1 if the checks are passed, 0 otherwise.
+*/
+static uint8_t sanity_checks(uint32_t addr, uint8_t *data, uint32_t length);
+
+/**
+ * @brief Return the minimum between two numbers.
+ * 
+ * The function uses signed integers in order to handle also negative numbers.
+ * 
+ * @param a first number.
+ * @param b second number.
+ * @return the minimum between a and b.
+*/
+static int32_t MIN(int32_t a, int32_t b) {
+    return (a < b) ? a : b;
+}
+
+
+/****************************************************************************/
+/**                                                                        **/
+/*                            GLOBAL VARIABLES                              */
+/**                                                                        **/
+/****************************************************************************/
+
+/**
+ * @brief SPI structure.
+*/
+spi_host_t spi;
+
+/**
+ * @brief Static vector used in the erase_and_write function.
+ * 
+ * It is used to store the data of a sector (4k) of the flash. Given the dimensions
+ * of the vector, is not possible to allocate it dinamically as it would cause a stack
+ * or heap overflow.
+*/
+uint8_t sector_data[4096];
+
+
+/****************************************************************************/
+/**                                                                        **/
+/*                           EXPORTED FUNCTIONS                             */
+/**                                                                        **/
+/****************************************************************************/
+uint8_t w25q128jw_init(spi_host_t spi_host) {
+    soc_ctrl_t soc_ctrl;
+    soc_ctrl.base_addr = mmio_region_from_addr((uintptr_t)SOC_CTRL_START_ADDRESS);
+
+    /*
+     * Check if memory mapped SPI is enabled. Current version of the bsp
+     * does not support memory mapped SPI.
+    */
+    if (get_spi_flash_mode(&soc_ctrl) == SOC_CTRL_SPI_FLASH_MODE_SPIMEMIO) {
+        return FLASH_ERROR; // Error
+    }
+
+    // Set the global spi variable to the one passed as argument.
+    spi = spi_host;
```

Comment:
why are you copying this?

Reply:
To make the spi base address a global variable in order to not pass it as a parameterer in every function of the bsp. I tought it was more elegant, but if it's bad practice I can remove it and make the spi a parameter to every function of the bsp.